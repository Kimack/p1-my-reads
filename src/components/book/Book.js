import React from 'react'
import Spinner from 'react-spinkit';
import PropTypes from 'prop-types';

class Book extends React.Component {

    state = {
        /**
         * Indicate if the item is still being processed
         */
        isUpdating: false
    }

    /**
     * @return the image url or empty if the image does not exist
     * TODO: Add placeholder image
     */
    getImageUrl = () => {
        const { book } = this.props;
        return book.imageLinks ? book.imageLinks.thumbnail : ""
    }

    /**
     * @return the current book shelf or none 
     */
    getBookShelf = () => {
        const { book } = this.props;
        return book.shelf ? book.shelf : "none";
    }

    /**
     * Called when the users wants to changed the book shelf
     */
    onBookShelfChange = (event) => {
        const { book, onBookShelfChange } = this.props;
        this.setState({ isUpdating: true });
        onBookShelfChange(book, event.target.value)
    }

    render() {
        const { book, shelves } = this.props;
        return (
            <div className="book">

                <div className="book-top">
                    {this.state.isUpdating ?
                        (
                            <div className="book-loading">
                                <Spinner className="books-loading-spinner" fadeIn="none" name="ball-pulse-sync" color="#60ac5d" />
                            </div>
                        ) :
                        (
                            <div className="book-shelf-changer">
                                <select onChange={this.onBookShelfChange} value={this.getBookShelf()}>
                                    <option value="none" disabled>Move to...</option>
                                    {shelves.map((shelf) => {
                                        return (<option key={shelf.id} value={shelf.id}>{shelf.title}</option>)
                                    })}
                                    <option value="none">None</option>
                                </select>
                            </div>
                        )}
                    <div className="book-cover" style={{ width: 128, height: 193, backgroundImage: `url("${this.getImageUrl()}")` }}></div>
                </div>
                <div className="book-title">{book.title}</div>
                <div className="book-authors">{book.authors}</div>
            </div>
        )
    }
}

Book.propTypes = {
	/**
	 * Current book
	 */
    book: PropTypes.object.isRequired,
	/**
	 * Available shelves array
	 */
    shelves: PropTypes.array.isRequired,
	/**
	 * Function that will be called when the shelf is changed
	 */
    onBookShelfChange: PropTypes.func.isRequired
}

export default Book;