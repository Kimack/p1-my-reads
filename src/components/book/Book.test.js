import React from 'react';
import { shallow, mount, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-15';
import Book from './Book';

const book = {
    title: "Title",
    authors: "Authors",
    imageLinks: { thumbnail: "imageUrl" },
    shelf: "none"
}

const shelves = [
    {
        id: "currentlyReading",
        title: "Currently Reading"
    },
    {
        id: "wantToRead",
        title: "Want to Read"
    },
    {
        id: "read",
        title: "Read"
    }
]

configure({ adapter: new Adapter() })


it('getImageUrl returns the correct value when not null', () => {
    const component = <Book book={{ imageLinks: { thumbnail: "imageUrl" } }} shelves={shelves} onBookShelfChange={() => { }} />;
    const test = mount(component);
    expect(test.instance().getImageUrl()).toBe("imageUrl");
});

it('getImageUrl returns "" when value is null', () => {
    const component = <Book book={{ imageLinks: null }} shelves={shelves} onBookShelfChange={() => { }} />;
    const test = mount(component);
    expect(test.instance().getImageUrl()).toBe("");
});

it('getBookShelf returns the correct value when not null', () => {
    const component = <Book book={{ shelf: "currentlyReading" }} shelves={shelves} onBookShelfChange={() => { }} />;
    const test = mount(component);
    expect(test.instance().getBookShelf()).toBe("currentlyReading");
});

it('getBookShelf returns "none" when value is null', () => {
    const component = <Book book={{ shelf: null }} shelves={shelves} onBookShelfChange={() => { }} />;
    const test = mount(component);
    expect(test.instance().getBookShelf()).toBe("none");
});

describe('book instance is rendered corretly', () => {
    it('renders without crashing', () => {
        const component = <Book book={book} shelves={shelves} onBookShelfChange={() => { }} />;
        expect(shallow(component)).toMatchSnapshot();
    });
});


describe('user click on shelf changed', () => {
    it('event onBookShelfChange is called', () => {
        const onBookShelfChange = jest.fn();
        const component = <Book book={book} shelves={shelves} onBookShelfChange={onBookShelfChange} />;
        const test = mount(component);
        test.find("select").simulate('change', { target: { value: 'currentlyReading' } });
        expect(onBookShelfChange).toHaveBeenCalledTimes(1);
        expect(onBookShelfChange).toBeCalledWith(book, 'currentlyReading');
    });

    it('state.loading is true and loading is rendered', () => {
        const onBookShelfChange = jest.fn();
        const component = <Book book={book} shelves={shelves} onBookShelfChange={onBookShelfChange} />;
        const test = mount(component);
        test.find("select").simulate('change', { target: { value: 'currentlyReading' } });
        expect(onBookShelfChange).toHaveBeenCalledTimes(1);
        expect(test.state("isUpdating")).toBeTruthy();
        expect(test).toMatchSnapshot();
    });

    it('state.loading is false after the component update', () => {
        const onBookShelfChange = jest.fn();
        const component = <Book book={book} shelves={shelves} onBookShelfChange={onBookShelfChange} />;
        let test = mount(component);

        //make the component state laoding
        test.setState({isUpdating:true});
        expect(test.state("isUpdating")).toBeTruthy();

        //send new props and check if the isUpdating changed to false
        test = test.setProps({book, shelves, onBookShelfChange});
        expect(test.state("isUpdating")).toBeFalsy();
        expect(test).toMatchSnapshot();
    });
});