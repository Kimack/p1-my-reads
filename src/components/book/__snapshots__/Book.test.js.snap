// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`book instance is rendered corretly renders without crashing 1`] = `
<div
  className="book"
>
  <div
    className="book-top"
  >
    <div
      className="book-shelf-changer"
    >
      <select
        onChange={[Function]}
        value="none"
      >
        <option
          disabled={true}
          value="none"
        >
          Move to...
        </option>
        <option
          key="currentlyReading"
          value="currentlyReading"
        >
          Currently Reading
        </option>
        <option
          key="wantToRead"
          value="wantToRead"
        >
          Want to Read
        </option>
        <option
          key="read"
          value="read"
        >
          Read
        </option>
        <option
          value="none"
        >
          None
        </option>
      </select>
    </div>
    <div
      className="book-cover"
      style={
        Object {
          "backgroundImage": "url(\\"imageUrl\\")",
          "height": 193,
          "width": 128,
        }
      }
    />
  </div>
  <div
    className="book-title"
  >
    Title
  </div>
  <div
    className="book-authors"
  >
    Authors
  </div>
</div>
`;

exports[`user click on shelf changed state.loading is false after the component update 1`] = `
<Book
  book={
    Object {
      "authors": "Authors",
      "imageLinks": Object {
        "thumbnail": "imageUrl",
      },
      "shelf": "none",
      "title": "Title",
    }
  }
  onBookShelfChange={[Function]}
  shelves={
    Array [
      Object {
        "id": "currentlyReading",
        "title": "Currently Reading",
      },
      Object {
        "id": "wantToRead",
        "title": "Want to Read",
      },
      Object {
        "id": "read",
        "title": "Read",
      },
    ]
  }
>
  <div
    className="book"
  >
    <div
      className="book-top"
    >
      <div
        className="book-shelf-changer"
      >
        <select
          onChange={[Function]}
          value="none"
        >
          <option
            disabled={true}
            value="none"
          >
            Move to...
          </option>
          <option
            key="currentlyReading"
            value="currentlyReading"
          >
            Currently Reading
          </option>
          <option
            key="wantToRead"
            value="wantToRead"
          >
            Want to Read
          </option>
          <option
            key="read"
            value="read"
          >
            Read
          </option>
          <option
            value="none"
          >
            None
          </option>
        </select>
      </div>
      <div
        className="book-cover"
        style={
          Object {
            "backgroundImage": "url(\\"imageUrl\\")",
            "height": 193,
            "width": 128,
          }
        }
      />
    </div>
    <div
      className="book-title"
    >
      Title
    </div>
    <div
      className="book-authors"
    >
      Authors
    </div>
  </div>
</Book>
`;

exports[`user click on shelf changed state.loading is true and loading is rendered 1`] = `
<Book
  book={
    Object {
      "authors": "Authors",
      "imageLinks": Object {
        "thumbnail": "imageUrl",
      },
      "shelf": "none",
      "title": "Title",
    }
  }
  onBookShelfChange={[Function]}
  shelves={
    Array [
      Object {
        "id": "currentlyReading",
        "title": "Currently Reading",
      },
      Object {
        "id": "wantToRead",
        "title": "Want to Read",
      },
      Object {
        "id": "read",
        "title": "Read",
      },
    ]
  }
>
  <div
    className="book"
  >
    <div
      className="book-top"
    >
      <div
        className="book-loading"
      >
        <Spinner
          className="books-loading-spinner"
          color="#60ac5d"
          fadeIn="none"
          name="ball-pulse-sync"
          noFadeIn={false}
          overrideSpinnerClassName=""
        >
          <div
            className="sk-spinner books-loading-spinner ball-pulse-sync"
            color="#60ac5d"
            style={
              Object {
                "color": "#60ac5d",
              }
            }
          >
            <div
              key="0"
            />
            <div
              key="1"
            />
            <div
              key="2"
            />
          </div>
        </Spinner>
      </div>
      <div
        className="book-cover"
        style={
          Object {
            "backgroundImage": "url(\\"imageUrl\\")",
            "height": 193,
            "width": 128,
          }
        }
      />
    </div>
    <div
      className="book-title"
    >
      Title
    </div>
    <div
      className="book-authors"
    >
      Authors
    </div>
  </div>
</Book>
`;
